Sources/NIOCore/Docs.docc/ByteBuffer-lengthPrefix.md

```diff
@@ -0,0 +1,119 @@
+strategies.
+
+## ``NIOBinaryIntegerEncodingStrategy`` protocol
+
+The first building block is a protocol which describes how to encode and decode an integer.
+
+An implementation of this protocol is needed for any encoding strategy. Currently, we provide one
+implementation: ``ByteBuffer/QUICBinaryEncodingStrategy``
+
+This protocol only has 2 requirements which don't have default implementations:
+
+- `readInteger`: Reads an integer from the ByteBuffer using this encoding. Implementations will read as many bytes as
+  they need to, according to their wire format, and move the reader index accordingly
+- `writeInteger`: Write an integer to the ByteBuffer using this encoding. Implementations will write as many bytes as
+  they need to, according to their wire format, and move the writer index accordingly.
+
+Note that implementations of this protocol need to either:
+
+- Encode the length of the integer into the integer itself when writing, so it knows how many bytes to read when
+  reading. This is what QUIC does.
+- Always use the same length, e.g. a simple strategy which always writes the integer as a UInt64.
+
+## Extensions on ``ByteBuffer``
+
+To provide a more user-friendly API, we have added extensions on `ByteBuffer` for writing integers with a
+chosen ``NIOBinaryIntegerEncodingStrategy``. These are ``ByteBuffer/writeEncodedInteger(_:strategy:)``
+and ``ByteBuffer/readEncodedInteger(_:as:)``.
+
+## Reading and writing length-prefixed data
+
+We added further APIs on ByteBuffer for reading data, strings and buffers which are written with a length prefix. These
+APIs first read an integer using a chosen encoding strategy. The integer then dictates how many bytes of data are read
+starting from after the integer.
+
+Similarly, there are APIs which take data, write its length using the provided strategy, and then write the data itself.
+
+## Writing complex data with a length-prefix
+
+Consider the scenario where we want to write multiple pieces of data with a length-prefix, but it is difficult or
+complex to work out the total length of that data.
+
+We decided to add the following API to ByteBuffer:
+
+```swift
+/// - Parameters:
+///     - strategy: The strategy to use for encoding the length.
+///     - writeData: A closure that takes a buffer, writes some data to it, and returns the number of bytes written.
+/// - Returns: Number of total bytes written. This is the length of the written data + the number of bytes used to write the length before it.
+public mutating func writeLengthPrefixed<Strategy: NIOBinaryIntegerEncodingStrategy>(
+    strategy: Strategy,
+    writeData: (_ buffer: inout ByteBuffer) throws -> Int
+) rethrows -> Int
+```
+
+Users could use the function as follows:
+
+```swift
+myBuffer.writeLengthPrefixed(strategy: .quic) { writer in
+    writer.writeString("something")
+    writer.writeSomethingComplex(something)
```

Comment:
nit: docs above use `buffer` as the name of the param passed to the closure. Seems odd to not use that here as well.